libcontainer: add support for Intel RDT/CAT in runc

xiaochenshen opened the pull request:
**v5 commits:**
4d2756c116315cd30b7ded82dfbb4bca78ceb627 libcontainer: add test cases for Intel RDT/CAT
692f6e1e27b2e4d38e1c9e0b463f7d1bfe3f7492 libcontainer: add support for Intel RDT/CAT in runc
af3b0d9dce34fb0e91c58d823d0ad48268b0f988 libcontainer/SPEC.md: add documentation for Intel RDT/CAT

**Changes in v5:**
Reworked code according to @crosbymichael and @hqhq's comments:
1. https://github.com/opencontainers/runc/pull/1279#issuecomment-319085278
2. https://github.com/opencontainers/runc/pull/1279#issuecomment-320008325
3. https://github.com/opencontainers/runc/pull/1279#issuecomment-319917615

- Obsoleted the design of "refactor resource manager interface": removed generic resourceManagers, isolated IntelRdt functions without touching cgroups code, removed the static map with static keys on resourceManagers set.
- Simplified IntelRdtManager implementation, removed "dead and empty" methods - GetPids(), GetAllPids(), Freeze(). Kept the required methods set - Apply(), GetStats(), Destroy(), GetPath(), and Set().
- Re-implemented GetStats(), isolated cgroups.Stats and intelrdt.Stats and kept type safety on the Stats.
- Changed test stuffs according to the functional rework.
- Added read-only L3 cache info (CbmMask, MinCbmBits, and NumClosids) into Intelrdt.Stats. Runc users could run "runc events" to fetch the information, which helps know more about L3 cache capabilities and how to set correct L3 cache schema in different Intel Xeon platforms.

---

**v4 commits**:
b1c836632b1d426a66b0e39abbac3df176177bd2 libcontainer: add support for Intel RDT/CAT in runc
48d8ffe6b84a8d647cf1622a0b02a544752ac791 libcontainer/SPEC.md: add documentation for Intel RDT/CAT
8851a0db77a65b42f72e6711356dd401204ad436 libcontainer: refactor resource manager interface
80648193d0add2b155844e558634b41eabb9116f vendor: specs-go: update specs for Intel RDT/CAT

**Changes in v4:**
Rebased to latest master branch.

---

**v3 commits**:
be25a2030bec8113434e53e0ee1b99a6e0bfc7ed libcontainer: add support for Intel RDT/CAT in runc
96fd05fc7ab0bcca414efc3e9e4894dbe1df0e5d libcontainer/SPEC.md: add documentation for Intel RDT/CAT
d5ac70d5ae68b514421a7c7f60475189191c70f1 libcontainer: refactor resource manager interface
c302b704bedd62e7655a401697f1a4a79085bcc0 vendor: specs-go: update specs for Intel RDT/CAT

**Changes in v3:**
Addressed comment https://github.com/opencontainers/runc/pull/1279#discussion_r116414850 from @yummypeng :
- Add parameter `path` to record IntelRdtPath in NewIntelRdtManager.

---

**v2 commits (for keeping records in code review):**
ab32d21313aa82dd6c88b24c1bfcbaf8a970dcea libcontainer: add support for Intel RDT/CAT in runc
5972fd56b5f59444a1a7260c84f915515622e297 libcontainer/SPEC.md: add documentation for Intel RDT/CAT
7f8b32107131150a581b62ad0f2665c8a03abbf1 libcontainer: refactor resource manager interface
9b03a51cbfaad7d0a2afe7c9667ac21d1799dcb0 vendor: specs-go: update specs for Intel RDT/CAT

**Changes in v2:**
Addressed the comments from @hqhq and @yummypeng :
- Added intelrdt config validator.
- Removed resctrl filesystem re-mount logic in isIntelRdtMounted().
- Optimized IsIntelRdtEnabled() to avoid unnecessary time-consuming mount points parsing.
- Changed Intel RDT related json items to snake_case style.
- Fixed some redundant codes with regard to intelRdtPath.
- Fixed potential null pointer issue in IntelRdtManager.Set().
- Fixed some typos (Sets, num_closids and etc.)
- Improved libcontainer/SPEC.md: added minimal supported kernel version, provided a more sophisticated config example.

---

**v1 commits (for keeping records in code review):**
4752fd2f639a38eb6ef6c2e030252098cd11d9f0 libcontainer: add support for Intel RDT/CAT in runc
c8486a7a8866b5ce4264ee0a17e54292a0f56fbf libcontainer/SPEC.md: add documentation for Intel RDT/CAT
aefd02ae210105573dabbb90ace3ab70943504ce libcontainer: refactor resource manager interface
8e8e5d96e6ad7dfefcd0f7d49a8d8c0b5cba0ef6 vendor: specs-go: update specs for Intel RDT/CAT

---

This PR fixes issue #433 

About Intel RDT/CAT feature:
Intel platforms with new Xeon CPU support Intel Resource Director Technology
(RDT). Cache Allocation Technology (CAT) is a sub-feature of RDT, which
currently supports L3 cache resource allocation.

This feature provides a way for the software to restrict cache allocation to a
defined 'subset' of L3 cache which may be overlapping with other 'subsets'.
The different subsets are identified by class of service (CLOS) and each CLOS
has a capacity bitmask (CBM).

For more information about Intel RDT/CAT can be found in the section 17.17
of Intel Software Developer Manual.

About Intel RDT/CAT kernel interface:
In Linux 4.10 kernel or newer, the interface is defined and exposed via
"resource control" filesystem, which is a "cgroup-like" interface.

Comparing with cgroups, it has similar process management lifecycle and
interfaces in a container. But unlike cgroups' hierarchy, it has single level
filesystem layout.

Intel RDT "resource control" filesystem hierarchy:
```
mount -t resctrl resctrl /sys/fs/resctrl
tree /sys/fs/resctrl
/sys/fs/resctrl/
|-- info
|   |-- L3
|       |-- cbm_mask
|       |-- min_cbm_bits
|       |-- num_closids
|-- cpus
|-- schemata
|-- tasks
|-- <container_id>
    |-- cpus
    |-- schemata
    |-- tasks
```
For runc, we can make use of `tasks` and `schemata` configuration for L3 cache
resource constraints.

The file `tasks` has a list of tasks that belongs to this group (e.g.,
<container_id>" group). Tasks can be added to a group by writing the task ID
to the "tasks" file  (which will automatically remove them from the previous
group to which they belonged). New tasks created by fork(2) and clone(2) are
added to the same group as their parent. If a pid is not in any sub group, it
Is in root group.

The file `schemata` has allocation bitmasks/values for L3 cache on each socket,
which contains L3 cache id and capacity bitmask (CBM).
```
	Format: "L3:<cache_id0>=<cbm0>;<cache_id1>=<cbm1>;..."
```
For example, on a two-socket machine, L3's schema line could be `L3:0=ff;1=c0`
which means L3 cache id 0's CBM is 0xff, and L3 cache id 1's CBM is 0xc0.

The valid L3 cache CBM is a *contiguous bits set* and number of bits that can
be set is less than the max bit. The max bits in the CBM is varied among
supported Intel Xeon platforms. In Intel RDT "resource control" filesystem
layout, the CBM in a group should be a subset of the CBM in root. Kernel will
check if it is valid when writing. e.g., 0xfffff in root indicates the max bits
of CBM is 20 bits, which mapping to entire L3 cache capacity. Some valid CBM
values to set in a group: 0xf, 0xf0, 0x3ff, 0x1f00 and etc.

For more information about Intel RDT/CAT kernel interface:  
https://www.kernel.org/doc/Documentation/x86/intel_rdt_ui.txt

An example for runc:
```
Consider a two-socket machine with two L3 caches where the default CBM is
0xfffff and the max CBM length is 20 bits. With this configuration, tasks
inside the container only have access to the "upper" 80% of L3 cache id 0 and
the "lower" 50% L3 cache id 1:

"linux": {
	"intelRdt": {
		"l3CacheSchema": "L3:0=ffff0;1=3ff"
	}
}
```
Signed-off-by: Xiaochen Shen <xiaochen.shen@intel.com>

xiaochenshen commented:
@cyphar @crosbymichael @hqhq @mrunalp @vishh
/cc @opencontainers/runc-maintainers

This PR will obsolete #1198 

To address @crosbymichael and @cyphar 's comments https://github.com/opencontainers/runc/pull/1198#issuecomment-263357031 and https://github.com/opencontainers/runc/pull/1198#issuecomment-263568672, the design is updated:

It adds a new "ResourceManager" structure as the base interface for all resource managers, such as cgroups manager and incoming IntelRdt manager.

All registered resource managers are consolidated in linuxContainer structure. We can apply to unified operations (e.g., Apply(), Set(), Destroy()) using all of the registered resource managers.

Currently, cgroups manager is the single resource manager in libcontainer. Linux kernel 4.10 will introduce Intel RDT/CAT feature, the kernel interface is exposed via "resource control" filesystem, which is a cgroup-like interface. In order to support Intel RDT/CAT in libcontainer, we need a new resource manager (IntelRdt manager) outside cgroups.

crosbymichael commented:
Thanks.  Design looks much better than before.

xiaochenshen commented:
@opencontainers/runc-maintainers
Any comments for this PR? Thank you.

xiaochenshen commented:
@LK4D4 @cyphar @crosbymichael @hqhq @mrunalp @vishh 
/cc @opencontainers/runc-maintainers

Who can help forward me the simplest steps to build runc in my forked repo (github.com/xiaochenshen/runc)? Thank you.

When I tried to rebase the code against latest runc master branch to resolve the conflict, I found the Godeps is replaced by vndr tool recently. I have tried to build with "make" in github.com/opencontainers/runc, it works. But in my forked dir, I got following errors. It looks like vndr related issues.
```
# make
go build -i -ldflags "-X main.gitCommit="291bf601105c97dc1aa631fdfb7fca63c947319b" -X main.version=1.0.0-rc2" -tags "seccomp" -o runc .
# github.com/xiaochenshen/runc
./restore.go:111: cannot use spec (type *"github.com/xiaochenshen/runc/vendor/github.com/opencontainers/runtime-spec/specs-go".Spec) as type *"github.com/opencontainers/runc/vendor/github.com/opencontainers/runtime-spec/specs-go".Spec in field value
./utils_linux.go:154: cannot use spec (type *"github.com/xiaochenshen/runc/vendor/github.com/opencontainers/runtime-spec/specs-go".Spec) as type *"github.com/opencontainers/runc/vendor/github.com/opencontainers/runtime-spec/specs-go".Spec in field value
make: *** [runc] Error 2
```

cyphar commented:
@xiaochenshen Make sure that this project is cloned properly inside your `GOPATH`. In particular you should have something like this set up:

```bash
% mkdir ~/yourgopath
% export GOPATH=~/yourgopath
% mkdir -p $GOPATH/src/github.com/opencontainers/runc
% git clone https://github.com/opencontainers/runc $GOPATH/src/github.com/opencontainers/runc
# ...
% cd $GOPATH/src/github.com/opencontainers/runc
% make
# should work
```

xiaochenshen commented:
@cyphar Thank you for your guide. It works with github.com/opencontainers/runc.
But this issue happens only in my forked repo github.com/xiaochenshen/runc which is also under $GOPATH ($GOPATH/src/github.com/xiaochenshen/runc).
Should I modify vendor.conf file?

Or git clone my repo `https://github.com/xiaochen/runc` into the directory $GOPATH/src/github.com/opencontainers/runc? It looks weird, to some extent:
```
git clone https://github.com/xiaochen/runc $GOPATH/src/github.com/opencontainers/runc
```

xiaochenshen commented:
@crosbymichael /cc @opencontainers/runc-maintainers

To address your suggestion in https://github.com/opencontainers/runtime-spec/pull/630#pullrequestreview-25906691:
`IntelRdt` struct has been moved out of `Resources` and inside `Linux` now.

This PR has been updated accordingly.

xiaochenshen commented:
ping @crosbymichael @cyphar @mrunalp @hqhq 
/cc @rjnagal @vmarmol @dqminh @avagin

https://github.com/opencontainers/runtime-spec/pull/630 + this PR to fix issue #433. And https://github.com/opencontainers/runtime-spec/pull/630 has been merged.
 
Could you help code review this PR at your convenience? Thank you!
I just rebased the code to fix conflict with rootless cgroup manager

xiaochenshen commented:
@yummypeng Thank you for your help on code review. I will check up the issues one by one later.

yummypeng commented:
I tried to add the configuration of intelRdt in my `config.json`, then run a container in an environment without support of intelRdt, and unexpectedly, the container was started successfully. Seems there needs an validation of `Spec` maybe in https://github.com/opencontainers/runc/blob/master/libcontainer/configs/validate/validator.go#L24.

yummypeng commented on libcontainer/intelrdt/intelrdt.go:
```diff
@@ -0,0 +1,556 @@
+	// e.g., 0xfffff in root indicates the max bits of CBM is 20 bits,
+	// which mapping to entire L3 cache capacity. Some valid CBM values
+	// to set in a group: 0xf, 0xf0, 0x3ff, 0x1f00 and etc.
+	l3CacheSchema := container.IntelRdt.L3CacheSchema
```
This requires a check if `container.IntelRdt == nil`, otherwise this will panic.

xiaochenshen replied:
Yes, you are right. I will add check in case of null pointer.

yummypeng commented:
Since `resctrl` is a "cgroup-like" interface, why not mount it into container like cgroup does?

xiaochenshen commented:
@yummypeng 
> Since resctrl is a "cgroup-like" interface, why not mount it into container like cgroup does?

There are discussions about that in obsolete #1198. You can refer to @crosbymichael and @cyphar 's comments: https://github.com/opencontainers/runc/pull/1198#issuecomment-263357031 and https://github.com/opencontainers/runc/pull/1198#issuecomment-263568672.

xiaochenshen commented on libcontainer/resourcemanager/resource_manager_linux.go:
```diff
@@ -0,0 +1,34 @@
+	// restore the object later
+	GetPaths() map[string]string
+
+	// Set the resource as configured
```
Thanks for finding the typo.

xiaochenshen commented on libcontainer/container_linux.go:
```diff
@@ -432,9 +451,14 @@ func (c *linuxContainer) newSetnsProcess(p *Process, cmd *exec.Cmd, parentPipe,
 	if err != nil {
 		return nil, err
 	}
+	intelRdtPath, err := intelrdt.GetIntelRdtPath(c.ID())
```
Yes, you are right. I will fix it.

xiaochenshen commented on libcontainer/container_linux_test.go:
```diff
@@ -184,6 +256,15 @@ func TestGetContainerState(t *testing.T) {
+		if path == "" {
+			t.Fatal("intel rdt path should not be empty")
+		}
+		if intelRdtPath := path; intelRdtPath != expectedIntelRdtPath {
```
Yes, I will fix it.

xiaochenshen commented on libcontainer/intelrdt/intelrdt.go:
```diff
@@ -0,0 +1,556 @@
+
+	// Returns Intel RDT "resource control" filesystem path to save in
+	// a state file and to be able to restore the object later
+	GetPath() string
```
Currently I kept GetPaths() here intentionally. In future, we may add more Intel RDT resource constrains besides L3 cache. If it is really not needed at that time, I will reorg the `Manager` interface just like you suggested.

xiaochenshen commented on libcontainer/intelrdt/intelrdt.go:
```diff
@@ -0,0 +1,556 @@
+			return false
+		}
+
+		// If not mounted, we try to mount again:
```
I agree with you now. I will remove this.

I was thinking that it is no harm for it. Unlike cgroups, Intel RDT resource control filesystem will likely not be mounted during boot-up by default in popular Linux OS distributions.

hqhq replied:
Shouldn't this be part of default mount info as we did for cgroups in https://github.com/opencontainers/runc/blob/master/libcontainer/specconv/example.go#L111-L116 and be handled as bind mount in [mountToRootfs](https://github.com/opencontainers/runc/blob/v1.0.0-rc3/libcontainer/rootfs_linux.go#L153)?

xiaochenshen replied:
@hqhq 
Unlike cgroup and cgroup filesystem, Intel RDT and resource control (resctrl) filesystem heavily depend on hardware/CPU support.

Likely, resctrl filesystem will not be mounted to rootfs either (1) h/w doesn't support Intel RDT or (2) kernel version older than 4.10, doesn't support Intel RDT.

hqhq replied:
I mean resctrl should also be bind mounted to container if we mount it on host by default, I didn't notice you're going to remove this, I'm OK with that, let users mount this on host and add that mount info to `config.json` specifically would be better.

xiaochenshen replied:
@hqhq 
> let users mount this on host and add that mount info to config.json specifically would be better.

I agree that, if Intel RDT is supported by h/w and kernel, it is user's responsibility to mount the resctrl filesystem in host, and then to add mount info to config.json accordingly.

In most cases (e.g., non-Intel Xeon platforms), Intel RDT is not supported and resctrl filesystem is not enabled. In my opinion, we may not add "fixed" rectrl filesystem mount info into specs.Mount in libcontainer/specconv/example.go by default.

xiaochenshen commented on libcontainer/intelrdt/intelrdt.go:
```diff
@@ -0,0 +1,556 @@
+	if err != nil {
+		return nil, err
+	}
+	numClosid, err := getIntelRdtParamUint(path, "num_closid")
```
Thanks for finding this typo.

xiaochenshen commented on libcontainer/intelrdt/intelrdt.go:
```diff
@@ -0,0 +1,556 @@
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	path, err := d.join(m.Id)
```
Do you mean the lock is not necessary or the lock here will cause a deadlock?

xiaochenshen replied:
@yummypeng 

For the nested containers case, I added comments in:
https://github.com/opencontainers/runc/pull/1279#discussion_r114473075

hqhq commented on libcontainer/resourcemanager/resource_manager_linux.go:
```diff
@@ -0,0 +1,34 @@
+// +build linux
```
Is this gonna break Solaris and Windows?

xiaochenshen replied:
@hqhq Which is better in your opinion? Thank you.
(1) Just remove this line (+// +build linux) in resource_manager_linux.go.
(2) Add new file resource_manager_unsupported.go for !linux cases:
```
// +build !linux

package resourcemanager
```

hqhq replied:
After a second check, I think current implementation is fine, it's only imported by packages which are wrapped by `linux` build flag, sorry for the noise.

hqhq commented on libcontainer/intelrdt/intelrdt.go:
```diff
@@ -0,0 +1,556 @@
+}
+
+func (raw *intelRdtData) join(id string) (string, error) {
+	path := filepath.Join(raw.root, id)
```
What about nested container cases? You can refer to the cgroup `join` function to see how cgroup handle this.

xiaochenshen replied:
@hqhq 

I was thinking if Intel RDT/CAT is possible to support nested containers case.

Unlike cgroups' hierarchy, Intel RDT resource control filesystem supports only single level filesystem layout (see details in section "Intel RDT "resource control" filesystem hierarchy" in this PR).

hqhq commented on events.go:
```diff
@@ -95,6 +96,12 @@ type memory struct {
+type intelRdt struct {
+	// The read-only default "schema" in root, for reference
+	L3CacheSchemaRoot string `json:"l3CacheSchemaRoot,omitempty"`
+	L3CacheSchema     string `json:"l3CacheSchema,omitempty"`
```
json variables should be snake_case.

xiaochenshen replied:
Got it. I will fix it.

hqhq commented on libcontainer/SPEC.md:
```diff
@@ -154,6 +154,89 @@ that no processes or threads escape the cgroups.  This sync is
+hardware and kernel support Intel RDT/CAT.
+
+In Linux kernel, it is exposed via "resource control" filesystem, which is a
+"cgroup-like" interface.
```
Can you specify the minimum linux kernel version that supports intelrdt feature?

xiaochenshen replied:
No problem - 4.10 kernel.

```+Linux kernel 4.10 introduces Intel RDT/CAT support, it is exposed via "resource control" filesystem...```

hqhq commented:
@xiaochenshen Thanks for working on this, overall it looks good, can you address all the comments and rebase please? thanks.

xiaochenshen commented on libcontainer/intelrdt/intelrdt.go:
```diff
@@ -0,0 +1,556 @@
+	isFlagSet, err := parseCpuInfoFile("/proc/cpuinfo")
+	if err != nil {
+		return false
+	}
```
Thanks, I agree with you.

xiaochenshen commented:
@yummypeng 
>I tried to add the configuration of intelRdt in my config.json, then run a container in an environment without support of intelRdt, and unexpectedly, the container was started successfully. Seems there needs an validation of Spec maybe in https://github.com/opencontainers/runc/blob/master/libcontainer/configs/validate/validator.go#L24.

Thank you. I will add Intel RDT configuration check to fix this issue.

xiaochenshen commented:
@hqhq 
> @xiaochenshen Thanks for working on this, overall it looks good, can you address all the comments and rebase please? thanks.

Thank you very much for helping code review. Most of the comments from you and @yummypeng will be addressed in my new patch.

But one remaining issue is nested containers case. Could you share some docs or issue/PR# about nested containers for me to learn from? I will figure out how to support it with single level resctrl filesystem interface and number of groups limitation. Thank you.

yummypeng commented:
@xiaochenshen Sorry, I didn't notice that Intel RDT resource control filesystem supports only single level filesystem layout, if it's true, then it seems that nested containers case cannot be supported. 

@hqhq Any ideas?

hqhq commented:
@xiaochenshen Nested container means you can create a container in another container, for cgroup, we bind mount the cgroup paths of current process to paths `/sys/fs/cgroup/<subsystem>` in container, which means if you create a container con2 in container con1, path `/sys/fs/cgroup/cpu/con2` in container con1 is actually `/sys/fs/cgroup/cpu/con1/con2` on host.

So if resctrl doesn't support:
```
tree /sys/fs/resctrl
/sys/fs/resctrl/
|-- info
|   |-- L3
|       |-- cbm_mask
|       |-- min_cbm_bits
|       |-- num_closids
|-- cpus
|-- schemata
|-- tasks
|-- <container_id1>
|   |-- <container_id2>
|       |-- cpus
|       |-- schemata
|       |-- tasks    
|   |-- cpus
|   |-- schemata
|   |-- tasks
``` 

or
```
mount -t resctrl resctrl /sys/fs/resctrl
unshare -m && chroot <container_rootfs>
mount -t resctrl resctrl /sys/fs/resctrl/
(the two mount points both works)
```

Then I think we can't support Intel RDT/CAT feature in nested containers.

xiaochenshen commented:
@hqhq @yummypeng 

Thanks for coaching me how nested containers work.

Intel RDT resource control filesystem doesn't support hierarchy layout which is the nature feature in cgroup.
Actually it is the reason why original cgroup-based Intel RDT/CAT kernel patch is rejected by kernel cgroup community. You can find more information from: https://github.com/opencontainers/runc/issues/433#issuecomment-281394321

xiaochenshen commented:
@hqhq @yummypeng 
@crosbymichael @cyphar
/cc @rjnagal @vmarmol @mrunalp @dqminh @avagin 

I have rebased the code to address the comments from @hqhq and @yummypeng.
Please help to review. Thank you.

**Changes in v2:**
Addressed the comments from @hqhq and @yummypeng:
- Added intelrdt config validator.
- Removed resctrl filesystem re-mount logic in isIntelRdtMounted().
- Optimized IsIntelRdtEnabled() to avoid unnecessary time-consuming mount points parsing.
- Changed Intel RDT related json items to snake_case style.
- Fixed some redundant codes with regard to intelRdtPath.
- Fixed potential null pointer issue in IntelRdtManager.Set().
- Fixed some typos (Sets, num_closids and etc.)
- Improved libcontainer/SPEC.md: added minimal supported kernel version, provided a more sophisticated config example.

**v1 commits for code review:**
4752fd2f639a38eb6ef6c2e030252098cd11d9f0 libcontainer: add support for Intel RDT/CAT in runc
c8486a7a8866b5ce4264ee0a17e54292a0f56fbf libcontainer/SPEC.md: add documentation for Intel RDT/CAT
aefd02ae210105573dabbb90ace3ab70943504ce libcontainer: refactor resource manager interface
8e8e5d96e6ad7dfefcd0f7d49a8d8c0b5cba0ef6 vendor: specs-go: update specs for Intel RDT/CAT

yummypeng commented on libcontainer/configs/validate/validator.go:
```diff
@@ -153,6 +157,19 @@ func (v *ConfigValidator) sysctl(config *configs.Config) error {
+		if !intelrdt.IsIntelRdtEnabled() {
+			return fmt.Errorf("intelRdt is specified in config, but Intel RDT is not supported")
+		}
+		if config.IntelRdt.L3CacheSchema == "" {
```
So, the following configuration is illegal?  
```
"linux": {
	"intelRdt": {
		"l3CacheSchema": ""
	}
}
```
@hqhq WDYT?

xiaochenshen replied:
@yummypeng 
> So, the following configuration is illegal?

In my opinion, if we specify "intelRdt" in config, we'd better specify a non-empty string for sub-item "l3CacheSchema".

ramkri123 commented:
Is this already committed? I tried generating a patch in the latest runc and it failed.

Feedback - would be very useful to have a method for specifying a percentage of the cache and abstract the low level cache bitmask details from the user.

Overall a very useful feature and well done.

xiaochenshen commented:
@rmanyari 
> Overall a very useful feature and well done.

Thank you for trying this and giving us positive feedback.

> Is this already committed? I tried generating a patch in the latest runc and it failed.

It is still under code review.

> Feedback - would be very useful to have a method for specifying a percentage of the cache and abstract the low level cache bitmask details from the user.

This PR is for basic Intel RDT/CAT feature support in runc. I do have plan to improve the user interface in future. Currently, users need to understand some background knowledge (e.g., the cache bitmask).

Specifying a percentage of the cache is really a good suggestion, but it also has limitation.
In this example, the contexts of "50% percentage" on cache 0 and on cache 1 are different:
```
On a two-socket machine with two L3 caches where the default CBM is 0xfffff and the
max CBM length is 20 bits. With this configuration, tasks inside the container only
have access to the "upper" 50% of L3 cache id 0 and the "lower" 50% L3 cache id 1:

"linux": {
	"intelRdt": {
		"l3CacheSchema": "L3:0=ffc00;1=3ff"
	}
}
```

ramkri123 commented:
@xiaochenshen  
>>> Specifying a percentage of the cache is really a good suggestion, but it also has limitation.
In this example, the contexts of "50% percentage" on cache 0 and on cache 1 are different:

Not really if this is the first partition you are creating ... at that point you have full control over the cache. Happy to chat more over a call as needed.

Additionally, you need a way to specify the exact partition number. Intel CAT allows several partitions.

xiaochenshen commented:
@ramkri123 

> Not really if this is the first partition you are creating ... at that point you have full control over the cache. Happy to chat more over a call as needed.

You can find more details about Intel RDT/CAT in the proposal https://github.com/opencontainers/runc/issues/433 and Intel RDT/CAT kernel interface:
https://www.kernel.org/doc/Documentation/x86/intel_rdt_ui.txt

> Additionally, you need a way to specify the exact partition number. Intel CAT allows several partitions.

Sorry, I didn't catch up with your question. I would like to explain more about how Intel RDT/CAT works in runc:

Intel RDT/CAT is exposed via "resource control" filesystem, which is a "cgroup-like" interface. Comparing with cgroups, it has similar process management lifecycle and interfaces in a container. But unlike cgroups' hierarchy, it has single level filesystem layout. After a runc container is started, the Intel RDT "resource control" filesystem hierarchy looks like:
```
tree /sys/fs/resctrl
/sys/fs/resctrl/
|-- info
|   |-- L3
|       |-- cbm_mask
|       |-- min_cbm_bits
|       |-- num_closids
|-- cpus
|-- schemata
|-- tasks
|-- <container_id>
    |-- cpus
    |-- schemata
    |-- tasks
```

/<container_id> is the group/partition for the runc container. The /<container_id>/schemata which specified via config.json is written by IntelRdtManager.Set(), and tasks running inside this container are attached to /<container_id>/tasks by IntelRdtManager.Apply() or WriteIntelRdtTasks().

ramkri123 commented:
@xiaochenshen 

Let us continue the discussion at opencontainers mailing list.

yummypeng commented on libcontainer/factory_linux.go:
```diff
@@ -224,10 +247,15 @@ func (l *LinuxFactory) Load(id string) (Container, error) {
+	resourceManagers := make(map[string]resourcemanager.ResourceManager)
+	resourceManagers["cgroups"] = l.NewCgroupsManager(state.Config.Cgroups, state.CgroupPaths)
+	if intelrdt.IsIntelRdtEnabled() {
+		resourceManagers["intelrdt"] = l.NewIntelRdtManager(&state.Config, id)
```
`state.IntelRdtPath` is ignored when load container from disk. This will result in residual directory in `/sys/fs/resctrl`.

Steps to reproduce:
```
# runc create testcontainer
# ls /sys/fs/resctrl
cpus  info  schemata  tasks  testcontainer
# runc delete testcontainer
# ls /sys/fs/resctrl
cpus  info  schemata  tasks  testcontainer
```

Seems that `NewIntelRdtManager` needs one more parameter to record `IntelRdtPath`.

xiaochenshen replied:
@yummypeng 

Thank you. I will add parameter `path` to record IntelRdtPath in NewIntelRdtManager.

xiaochenshen commented:
@yummypeng @hqhq 
@crosbymichael @cyphar
/cc @rjnagal @vmarmol @mrunalp @dqminh @avagin 

I have rebased the code to v3. Please help code review. Thank you.

**v3 commits**:
be25a2030bec8113434e53e0ee1b99a6e0bfc7ed libcontainer: add support for Intel RDT/CAT in runc
96fd05fc7ab0bcca414efc3e9e4894dbe1df0e5d libcontainer/SPEC.md: add documentation for Intel RDT/CAT
d5ac70d5ae68b514421a7c7f60475189191c70f1 libcontainer: refactor resource manager interface
c302b704bedd62e7655a401697f1a4a79085bcc0 vendor: specs-go: update specs for Intel RDT/CAT

**Changes in v3:**
Addressed comment https://github.com/opencontainers/runc/pull/1279#discussion_r116414850 from @yummypeng :
- Add parameter `path` to record IntelRdtPath in NewIntelRdtManager.

xiaochenshen commented:
@yummypeng @hqhq 
@crosbymichael @cyphar 
/cc @rjnagal @vmarmol @mrunalp @dqminh @avagin 

I have rebased the code against latest master branch. Please help code review at your convenience.
Thank you.

xiaochenshen commented:
Ping maintainers and contributors:

@hqhq @crosbymichael @cyphar @yummypeng 
/cc @rjnagal @vmarmol @mrunalp @dqminh @avagin 


Could you help code review at your convenience?
Thank you.

hqhq commented:
LGTM

@xiaochenshen Thanks for working on this, vendor probably needs rebase but github didn't complain so far.

hqhq commented:
@crosbymichael @cyphar I'm adding this to 1.0 milestone as it's part of runtime-spec 1.0.

crosbymichael commented:
Either things change a lot from the last time I looked at the design or I spaced it.

I still don't see why we have to try to fit this behind the cgroup manager interface.  It adds so much code, dead code, empty methods, etc that it should be clear that it probably is not the right place for this code.  Now we have a static map with static keys on a map and removed type safety on the stats.

I don't understand why this feature wasn't just a simple `SetRdt()` and `StatRdt()`.  

@cyphar what do you think?

hqhq commented:
@xiaochenshen Does intelRdt need the whole cgroup manager interface?

xiaochenshen commented:
@crosbymichael 
/cc @cyphar 
The general design has not changed from https://github.com/opencontainers/runc/pull/1279#issuecomment-273307523 which you have commented positive feeback in https://github.com/opencontainers/runc/pull/1279#issuecomment-277826103
```
It adds a new "ResourceManager" structure as the base interface for all resource managers,
such as cgroups manager and incoming IntelRdt manager.

All registered resource managers are consolidated in linuxContainer structure.
We can apply to unified operations (e.g., Apply(), Set(), Destroy()) using all of the
registered resource managers.

Currently, cgroups manager is the single resource manager in libcontainer.
Linux kernel 4.10 will introduce Intel RDT/CAT feature, the kernel interface is exposed
via "resource control" filesystem, which is a cgroup-like interface. In order to support
Intel RDT/CAT in libcontainer, we need a new resource manager (IntelRdt manager)
outside cgroups."
```

This design is also based on @cyphar 's suggestion in https://github.com/opencontainers/runc/issues/433#issuecomment-241212914
```
If we can't consolidate it with the cgroup code, then please make a new "resource manager"
structure where all Managers are consolidated. Then, we can just have generic Manager
code which applies to right operation using all of the registered managers. That way,
we don't have intelRdtWhatever sitting anywhere outside the "resource manager" code.
This would actually be a possible improvement over the current state of affairs (though
there's some cgroup code that you won't be able to refactor so this may result in the
patch being a bit harder to implement).
```

In this PR, commit 8851a0db77a65b42f72e6711356dd401204ad436 refactors current resource manager interface.
``` 
This patch adds a new "ResourceManager" structure as the base interface for
all resource managers, such as cgroups manager and incoming IntelRdt manager.

All registered resource managers are consolidated in linuxContainer structure.
We can apply to unified operations (e.g., Apply(), Set(), Destroy()) using
all of the registered resource managers.
```

As a result, IntelRdt manager doesn't depends on cgroups manager, IntelRdt manager and cgroups manager have same base interface "ResourceManager".

xiaochenshen commented:
@hqhq 
Thank you for code review on several iterations.

> @xiaochenshen Does intelRdt need the whole cgroup manager interface?

With this PR, IntelRdt manager doesn't depends on cgroups manager, IntelRdt manager and cgroups manager have same base interface "ResourceManager".

hqhq commented:
>> @xiaochenshen Does intelRdt need the whole cgroup manager interface?
>
> With this PR, IntelRdt manager doesn't depends on cgroups manager, IntelRdt manager and cgroups manager have same base interface "ResourceManager".

My point is you made "ResourceManager" the same as the old cgroup manager interface, so you have to implement all these functions in intelRdt manager, are these functions all needed by intelRdt which make them the base ResourceManager? Maybe that's "dead code, empty methods" which @crosbymichael mentioned.

Maybe you can have a subset of old cgroup manager functions as `ResourceManager` interface and have cgroup and intelRdt there own functions, so you don't have to implement functions you don't need such as `Freeze` and handle `GetStats` differently.

crosbymichael commented:
Yes, my point is that this is so much code to force this features to "be like cgroups" when its not.  We could do the same thing with two function calls in the libcontainer codebase to set and get this.  Its one field on the config struct and 1000+ lines to make it look like a cgroup.  I think we just take the simple route and just get the feature working without touching cgroups code.

xiaochenshen commented:
@crosbymichael @hqhq 
Thank you for your code review and comments.
I will try to re-work to align with your suggestions.

xiaochenshen commented:
@crosbymichael @hqhq
/cc @cyphar @yummypeng
/cc @opencontainers/runc-maintainers

## Design updates (2017-08-18)

**v5 commits:**
5f47f99f3016027ddb0ce28d811c56c16cbdc270 libcontainer: add test cases for Intel RDT/CAT
0e83c867e7f0febe09eae3dc6d5fb6ab7e5e3a52 libcontainer: add support for Intel RDT/CAT in runc
657b33c724fc1f302aad5c05aa448319009f9df4 libcontainer/SPEC.md: add documentation for Intel RDT/CAT

**Changes in v5:**
Reworked code according to @crosbymichael and @hqhq's comments:
1. https://github.com/opencontainers/runc/pull/1279#issuecomment-319085278
2. https://github.com/opencontainers/runc/pull/1279#issuecomment-320008325
3. https://github.com/opencontainers/runc/pull/1279#issuecomment-319917615

- Obsoleted the design of "refactor resource manager interface": removed generic resourceManagers, isolated IntelRdt functions without touching cgroups code, removed the static map with static keys on resourceManagers set.
- Simplified IntelRdtManager implementation, removed "dead and empty" methods - GetPids(), GetAllPids(), Freeze(). Kept the required methods set - Apply(), GetStats(), Destroy(), GetPath(), and Set().
- Re-implemented GetStats(), isolated cgroups.Stats and intelrdt.Stats and kept type safety on the Stats.
- Changed test stuffs according to the functional rework.
- Added read-only L3 cache info (CbmMask, MinCbmBits, and NumClosids) into Intelrdt.Stats. Runc users could run "runc events" to fetch the information, which helps know more about L3 cache capabilities and how to set correct L3 cache schema in different Intel Xeon platforms.

xiaochenshen commented:
@crosbymichael @hqhq 
/cc @cyphar @yummypeng 
/cc @opencontainers/runc-maintainers

Could you help code review at your convenience?
Thank you.

## Design updates (2017-08-18)

**v5 commits:**
5f47f99f3016027ddb0ce28d811c56c16cbdc270 libcontainer: add test cases for Intel RDT/CAT
0e83c867e7f0febe09eae3dc6d5fb6ab7e5e3a52 libcontainer: add support for Intel RDT/CAT in runc
657b33c724fc1f302aad5c05aa448319009f9df4 libcontainer/SPEC.md: add documentation for Intel RDT/CAT

**Changes in v5:**
Reworked code according to @crosbymichael and @hqhq's comments:
1. https://github.com/opencontainers/runc/pull/1279#issuecomment-319085278
2. https://github.com/opencontainers/runc/pull/1279#issuecomment-320008325
3. https://github.com/opencontainers/runc/pull/1279#issuecomment-319917615

- Obsoleted the design of "refactor resource manager interface": removed generic resourceManagers, isolated IntelRdt functions without touching cgroups code, removed the static map with static keys on resourceManagers set.
- Simplified IntelRdtManager implementation, removed "dead and empty" methods - GetPids(), GetAllPids(), Freeze(). Kept the required methods set - Apply(), GetStats(), Destroy(), GetPath(), and Set().
- Re-implemented GetStats(), isolated cgroups.Stats and intelrdt.Stats and kept type safety on the Stats.
- Changed test stuffs according to the functional rework.
- Added read-only L3 cache info (CbmMask, MinCbmBits, and NumClosids) into Intelrdt.Stats. Runc users could run "runc events" to fetch the information, which helps know more about L3 cache capabilities and how to set correct L3 cache schema in different Intel Xeon platforms.

crosbymichael commented on libcontainer/intelrdt/intelrdt.go:
```diff
@@ -0,0 +1,545 @@
+// Check if Intel RDT is enabled
+func IsIntelRdtEnabled() bool {
+	// We have checked the flag before
+	if isIntelRdtEnabled {
```
isn't this racy? you should probably use a `sync.Once` for initialization of this bool

xiaochenshen replied:
@crosbymichael 
Thank you for good suggestion. I will fix it in follow-up PR.

xiaochenshen replied:
I opened a new PR in https://github.com/opencontainers/runc/pull/1589 to fix this issue.

crosbymichael commented:
LGTM

Can you fix my comment in a follow up PR?

mrunalp commented:
LGTM. We can fix up smaller issues in follow-on PRs. Thanks!

mrunalp commented:
LGTM

xiaochenshen commented:
@crosbymichael @mrunalp 

Thank you for your code review.
I will open a new PR to fix remaining issues.

xiaochenshen commented:
@crosbymichael @hqhq @cyphar @mrunalp 
/cc @opencontainers/runc-maintainers @yummypeng

I planned to add Intel RDT/CAT support for Docker. More information can be found in #433 .
But I am confused with the "vendor" configuration among Docker related projects (containerd, moby and docker-cli). I found a lot of dependency issues because recent runc changes have not been merged into these projects yet.

Who could share me the tips or documents about "vendor" tools?
Thank you in advance.

cyphar commented:
There is a `vendor.conf` in the root of Docker's source tree. You need to just update the commit ID in that file, and then run `vndr github.com/opencontainers/runc` in the source tree (the tool is available at https://github.com/lk4d4/vndr).

xiaochenshen commented:
@cyphar 
Got it. Thank you.
